lib/stdlib/src/json.erl

```diff
@@ -0,0 +1,765 @@
+-type array_start_fun() :: fun((Acc :: dynamic()) -> ArrayAcc :: dynamic()).
+-type array_push_fun() :: fun((Value :: dynamic(), Acc :: dynamic()) -> NewAcc :: dynamic()).
+-type array_finish_fun() :: fun((ArrayAcc :: dynamic()) -> dynamic()).
+-type object_start_fun() :: fun((Acc :: dynamic()) -> ObjectAcc :: dynamic()).
+-type object_push_fun() :: fun((Key :: dynamic(), Value :: dynamic(), Acc :: dynamic()) -> NewAcc :: dynamic()).
+-type object_finish_fun() :: fun((ObjectAcc :: dynamic()) -> dynamic()).
+
+-type decoders() :: #{
+    empty_array => term(),
+    array_start => array_start_fun(),
+    array_push => array_push_fun(),
+    array_finish => array_finish_fun(),
+    empty_object => term(),
+    object_start => object_start_fun(),
+    object_push => object_push_fun(),
+    object_finish => object_finish_fun(),
+    float => from_binary_fun(),
+    integer => from_binary_fun(),
+    string => from_binary_fun(),
+    null => term()
+}.
+
+-record(decode, {
+    empty_array = [] :: term(),
+    array_start :: array_start_fun() | undefined,
+    array_push :: array_push_fun() | undefined,
+    array_finish = fun lists:reverse/1 :: array_finish_fun(),
+    empty_object = #{} :: term(),
+    object_start :: object_start_fun() | undefined,
+    object_push :: object_push_fun() | undefined,
+    object_finish = fun maps:from_list/1 :: object_finish_fun(),
+    float = fun erlang:binary_to_float/1 :: from_binary_fun(),
+    integer = fun erlang:binary_to_integer/1 :: from_binary_fun(),
+    string :: from_binary_fun() | undefined,
+    null = null :: term()
+}).
+
+-type acc() :: dynamic().
+-type stack() :: [?ARRAY | ?OBJECT | binary() | acc()].
+-type decode() :: #decode{}.
+
+-type simple_decode_value() ::
+    integer() |
+    float() |
+    boolean() |
+    null |
+    binary() |
+    list(simple_decode_value()) |
+    #{binary() => simple_decode_value()}.
+
+-doc "
+Parses a JSON value from `Binary`.
+
+Supports basic data mapping:
+| **JSON** | **Erlang**           |
+|----------|----------------------|
+| Number   | integer() \| float() |
+| Boolean  | true \| false        |
+| Null     | null                 |
+| String   | binary()             |
+| Object   | #{binary() => _}     |
+".
+-spec decode(binary()) -> simple_decode_value().
+decode(Binary) when is_binary(Binary) ->
+    case value(Binary, Binary, 0, ok, [], #decode{}) of
+        {Result, _Acc, <<>>} -> Result;
+        {_, _, Rest} -> unexpected(Rest, 0)
+    end.
+
+-doc "
+Parses a JSON value from `Binary`.
+
+Similar to `decode/1` escept the decoding process
+can be customized with the callbacks specifified in
+`Decoders`. The callbacks will use the `Acc` value
+as the initial accumulator.
+
+Any left-over, unparsed data in `Binary` will be returned.
+
+## Example
+
+Decoding object keys as atoms:
+
+```erlang
+Push = fun(Key, Value, Acc) -> [{binary_to_existing_atom(Key), Value} | Acc] end,
+json:decode(Data, #{object_push => Push})
+```
```

Comment:
There is no `json:decode/2`.
This example (and default code) also assumes the *Acc* is a list, should probably be documented.